### ISSUE TITLE
Code Explorer's Sort functionality is broken

### ISSUE THREAD

**Issue by ghost** [rubberduck-vba/Rubberduck#3307]
Consider a VB standard module that has the following methods contained in it, in the following order:

`Get()`
`Send()`
`Private DeleteIfNotSomething()`
`GoTo()`
`Add()`
`Delete()`

Now, consider the CE Sort menu:
![image](https://user-images.githubusercontent.com/28154402/29783174-0e7a88d0-8be5-11e7-88e1-057b63f61e1c.png)
It has what appears to be two sort options, "By Name" and "As in module". Additionally, there is a third option which appears to be a condition of the first two options, "[Group] by type".

This presents a combination of three choices to the user:
1) Sort by name
2) Group by type, then sort by name
3) As in Module (i.e., no sorting at all)
4) Group by type, then display as in Module.

Sorting works as expected, except for combination 4.  It groups items by Type, but then fails to display the items in each group in the same order as occurs in the spreadsheet.  For example, the resultant list might be:
`Private DeleteIfNotSomething()`
`GoTo()`
`Add()`
`Get()`
`Send()`
`Delete()`

The code for doing the sort is:
```
        public void ReorderItems(bool sortByName, bool sortByType)
        {
            if (sortByType)
            {
                Items = sortByName
                    ? Items.OrderBy(o => o, new CompareByType()).ThenBy(t => t, new CompareByName()).ToList()
                    : Items.OrderBy(o => o, new CompareByType()).ThenBy(t => t, new CompareBySelection()).ToList();

                return;
            }

            Items = sortByName
                ? Items.OrderBy(t => t, new CompareByName()).ToList()
                : Items.OrderBy(t => t, new CompareBySelection()).ToList();
        }
```

The line responsible is:
`Items.OrderBy(o => o, new CompareByType()).ThenBy(t => t, new CompareBySelection()).ToList();`

However, sorting by As In Module works:
'Items.OrderBy(t => t, new CompareBySelection()).ToList();`

So, performing the OrderBy(Type) rearranges the list being passed to ThenBy(Selection).  or `CompareByType` is rearranging the list.

Thoughts or ideas?

**Comment by retailcoder** [comment]
Seems the combined sort needs its own comparer implementation

**Comment by shadowofsilicon** [comment]
@retailcoder You think that the current comparer implementation is rearranging the list?

**Comment by Vogel612** [comment]
FWIW it seems like a good idea to overhaul that method a bit and separate the assembling of comparators from the sorting.

**Comment by shadowofsilicon** [comment]
I have this figured out a bit.

It is sorting by *location* correctly.  However, before that, it sorts by *type* and it is treating A) `Sub` and `Function` as a separate type, and for each of those, it is sorting `Private` different than `Public` than `Implicit`.

However, in all of those combinations, the only two that have different icons and therefore are **visually** different in the context of the CE is `Private` and `Public`.

So, changing the compare algo to treat `Implicit` the same as `Public` seems like an obvious fix.

IMHO, `Subs` and `Functions` should also be grouped together, but I want to hear from the group before moving forward.

**Comment by comintern** [comment]
Closed via #4661
